src/bin/nydus-image/builder/directory.rs

```diff
@@ -124,13 +124,17 @@ impl Builder for DirectoryBuilder {
         // Scan source directory to build upper layer tree.
         let layer_idx = if bootstrap_ctx.layered { 1u16 } else { 0u16 };
         let mut tree = self.build_tree_from_fs(ctx, &mut bootstrap_ctx, layer_idx)?;
+        let origin_bootstarp_offset = bootstrap_ctx.offset;
         let mut bootstrap = Bootstrap::new()?;
         if bootstrap_ctx.layered {
             // Merge with lower layer if there's one, do not prepare `prefetch` list during merging.
             ctx.prefetch.disable();
             bootstrap.build(ctx, &mut bootstrap_ctx, &mut tree)?;
             tree = bootstrap.apply(ctx, &mut bootstrap_ctx, bootstrap_mgr, blob_mgr, None)?;
         }
+        // If layered, the bootstrap_ctx.offset will be set in first build, so we need restore it here
+        bootstrap_ctx.offset = origin_bootstarp_offset;
+        bootstrap_ctx.layered = false;
```

Comment:
This also affects https://github.com/dragonflyoss/image-service/blob/fdee5992103916910dea755a2df3d5d669ea6efb/src/bin/nydus-image/core/bootstrap.rs#L201

where `layered` is used to check whiteout, have you checked if setting `layered = false` would cause any difference?